conda_lock/lockfile/__init__.py

```diff
@@ -111,27 +138,31 @@ def dep_name(manager: str, dep: str) -> str:
 
         by_category[request.category].append(request.name)
 
-    # now, map each package to its root request preferring the ones earlier in the
-    # list
+    # now, map each package to every root request that requires it
     categories = [*categories, *(k for k in by_category if k not in categories)]
-    root_requests = {}
+    root_requests: DefaultDict[str, List[str]] = defaultdict(list)
```

Comment:
Before, we used to map each package (LockedDependency) to 1 root dependency that is its ancestor (Dependency). A root dependency is a dependency specified in a source file. So the root dependency uses the package as a sub-dependency. As the comment says, we would try to use the first root.

Now, we need to store all root dependencies, in case there are multiple that use a package as a sub-dependency. That will allow us to take multiple `category` values for each root and combine to create the `categories` value for a package.

Reply:
This is the core change for the second commit.